content/post/095-road-to-dht.md

```diff
@@ -0,0 +1,84 @@
+
+The IPFS DHT distributes and stores these pairs of keys and values in small tables across many nodes or peers throughout the network. Peers, in turn, store information about where to physically locate specific content. Kademlia, our DHT implementation, helps keep track of which nodes are _providing_ what data to others in the network.
+
+In theory, the DHT should make finding, providing, and fetching data among nodes a well-orchestrated and very efficient affair. However, throw in real world conditions and a rapidly scaling network, and things can go awry.
+
+## Let’s rewind
+
+The IPFS Public Network experienced substantial growth last year, quickly scaling 30x, with **hundreds of thousands of nodes** now participating in the network! Major software deployments and new decentralized apps came on board with expectations for a production-ready network with the reliability and performance to match. Many were using IPFS as a decentralized CDN (Content Delivery Network) to share and find content across the network. Unfortunately, due to a proliferation of new nodes unable to route peers to the desired content, the network was letting some of these users down.
+
+{{< youtube jpQnQbfhuBc >}}
+
+In early 2019, we started to hear rumblings around content routing issues, which were exacerbated as more and more nodes joined the network over Q1 and Q2 - many of them behind home firewalls or NATs. We were constantly bootstrapping and patching the network to remedy performance and reliability issues. But, we couldn’t really test the efficacy (or truly diagnose any regressions) of these patches until they were _in production_, making it hard to iterate and try out larger fixes to network configuration.
+
+We clearly had a problem, and we were hearing about it from the community. Our first step was to investigate thoroughly to determine what fixes we should prioritize first, and identify the tooling we needed to land our changes successfully.  
+
+## Narrowing in
+
+After some investigation, it was clear that our DHT implementation, combined with the recent growth of undialable nodes, was the culprit behind the slow, unreliable content routing. We had a chance to investigate this more deeply while we were together at IPFS Camp, and discovered a few problem areas:
+
+* **Peer Availability:** As new peers joined the network we treated them equally, but most couldn’t actually be reached because they were behind firewalls or NATs. DHT query time was wasted trying to dial these peers that never got requests.
+* **Query Termination:** Once a DHT query came upon the data it needed, it didn’t stop! It continued searching until it found all the closest peers.
+* **Routing Table Maintenance:** Routing tables were clogged with many undialable peers, resulting in searches that were linear in the number of peers they queried instead of the expected log-scale efficiency.
```

Comment:
@Stebalien the non O(log_k(N)) lookups come from a combination of Query Termination and Routing Table Maintenance. Do you think it makes sense to just leave it where it is, move it to the query termination bullet, or scrap it and replace with something like "Routing tables were clogged with many undialable peers, resulting in searches that did not progress as quickly as would be expected with the Kademlia protocol"

Reply:
I think it's fine

Reply:
It's a little more complicated. The main issue here was that many routing tables weren't full enough. That is, peers didn't actively search to fill up their routing table buckets.